src/shims/time.rs

```diff
@@ -0,0 +1,117 @@
+    // `ImmTy::from_int` panic.
+    let size = layout.size;
+    let truncated = truncate(int as u128, size);
+    if sign_extend(truncated, size) as i128 != int {
+        throw_unsup_format!(
+            "Signed value {:#x} does not fit in {} bits",
+            int,
+            size.bits()
+        )
+    }
+    Ok(ImmTy::from_int(int, layout))
+}
+
+impl<'mir, 'tcx> EvalContextExt<'mir, 'tcx> for crate::MiriEvalContext<'mir, 'tcx> {}
+pub trait EvalContextExt<'mir, 'tcx: 'mir>: crate::MiriEvalContextExt<'mir, 'tcx> {
+    // Foreign function used by linux
+    fn clock_gettime(
+        &mut self,
+        clk_id_op: OpTy<'tcx, Tag>,
+        tp_op: OpTy<'tcx, Tag>,
+    ) -> InterpResult<'tcx, i32> {
+        let this = self.eval_context_mut();
+
+        if !this.machine.communicate {
+            throw_unsup_format!("`clock_gettime` not available when isolation is enabled")
+        }
+
+        let clk_id = this.read_scalar(clk_id_op)?.to_i32()?;
+        if clk_id != this.eval_libc_i32("CLOCK_REALTIME")? {
+            let einval = this.eval_libc("EINVAL")?;
+            this.set_last_error(einval)?;
+            return Ok(-1);
+        }
+
+        let tp = this.force_ptr(this.read_scalar(tp_op)?.not_undef()?)?;
+
+        let (duration, sign) = get_time();
+        let tv_sec = sign * (duration.as_secs() as i128);
+        let mut tv_nsec = duration.subsec_nanos() as i128;
+        // If the number of seconds is zero, we need to put the sign into the second's fraction.
+        if tv_sec == 0 && sign < 0 {
+            tv_nsec *= sign;
+        }
+
+        let imms = [
+            int_to_immty_checked(tv_sec, this.libc_ty_layout("time_t")?)?,
+            int_to_immty_checked(tv_nsec, this.libc_ty_layout("c_long")?)?,
+        ];
+
+        this.write_immediates(&tp, &imms)?;
+
+        Ok(0)
+    }
+    // Foreign function used by generic unix
+    fn gettimeofday(
+        &mut self,
+        tv_op: OpTy<'tcx, Tag>,
+        tz_op: OpTy<'tcx, Tag>,
+    ) -> InterpResult<'tcx, i32> {
+        let this = self.eval_context_mut();
+
+        if !this.machine.communicate {
+            throw_unsup_format!("`gettimeofday` not available when isolation is enabled")
+        }
+        // Using tz is obsolete and should always be null
+        let tz = this.read_scalar(tz_op)?.not_undef()?;
+        if !this.is_null(tz)? {
+            let einval = this.eval_libc("EINVAL")?;
+            this.set_last_error(einval)?;
+            return Ok(-1);
+        }
+
+        let tv = this.force_ptr(this.read_scalar(tv_op)?.not_undef()?)?;
+
+        let (duration, sign) = get_time();
+        let tv_sec = sign * (duration.as_secs() as i128);
+        let mut tv_usec = duration.subsec_micros() as i128;
+        // If the number of seconds is zero, we need to put the sign into the second's fraction.
+        if tv_sec == 0 && sign < 0 {
+            tv_usec *= sign;
+        }
```

Comment:
Same deal with negative times, do you have docs that confirm that these are supported and how they are represented?

Reply:
No I don't have any docs, but those integers are signed so I assumed incorrectly that negative values were possible. After digging a little bit it seems that in fact it is undesired to return negative times.

What should be an appropiate error for this? According to `man`, this function only sets the last error for other kinds of undesired behaviors. Should I throw an unsupported error with a "Time went backwards" message?

Reply:
I'd `throw_unsup` when detecting a negative time.